### DISCUSSION TITLE
第20回 PHASE/0 利用講習会の準備と反省

### DISCUSSION THREAD

**Discussion by matelier** [Materials-Science-Software-Consortium/phase0_tutorial_wsl#25] [General]
# 開催概要

https://azuma.nims.go.jp/events/semi2024/semi20240830
2024年8月30日(金)
東京大学 生産技術研究所

**Comment by matelier** [comment]
# 準備編

講習会で利用するPHASE/0のバージョンを決めたい。

提案：`2024.01`を使いましょう。

（基礎編の）講習会は、基本的な機能のみを利用するので、PHASE/0のバージョンは関係なさそうに思えますが、（`2023.01.01`でない）`2023.01`は、gfortranでコンパイルすると誘電関数が計算できない、ことが知られています。

`2023.01`は講習会に不向きです。

また、講習会で利用する可能性のあるスクリプトには、`phase`バイナリの位置がディレクトリ含めて記述されています。
https://github.com/Materials-Science-Software-Consortium/phase0_tutorial_wsl/blob/main/wsl/Si2/ev/cubic.sh

テキスト（説明資料）は、バージョン番号を読み替えて利用することが難しくありませんが、スクリプトの書き換えは（Linuxでのテキストエディタの扱いに不慣れな）受講者には負担になる可能性があります。
https://phase0-tutorial-wsl.readthedocs.io/ja/latest/

過去の反省
https://github.com/Materials-Science-Software-Consortium/phase0_tutorial_wsl/discussions/21#discussion-5561369

**Reply by matelier** [reply]
https://phase0-tutorial-wsl.readthedocs.io/ja/latest/intro.html#phase-0

にて、PHASE/0論文が紹介されています。
本チュートリアルは誘電関数計算を含みますので、UVSOR論文も並列して挙げた方が良くないでしょうか。
（もしくは、両論文共挙げない）

--------------------------

論文よりも、マニュアルへのリンクの方が重要度が高いと思います。
https://phase0.readthedocs.io/

**Reply by matelier** [reply]
（講習会とは関係ない話題です；macOS向け）

[概要](https://phase0-tutorial-wsl.readthedocs.io/ja/latest/intro.html#id2)では、

> 一般的なGNU/LinuxやmacOSでも実行できるはずです。

とありますが、EV曲線の例題は、macOSでこの通りに実行しても、

https://phase0-tutorial-wsl.readthedocs.io/ja/latest/si.html#id43

期待した結果にならないことを、かつてコメントしました。

https://github.com/Materials-Science-Software-Consortium/phase0_tutorial_wsl/issues/3#issuecomment-1509499800

**Reply by matelier** [reply]
（workfuncのコンパイルについて（に限らず）、すでにDiscussionで述べたこと#23 は繰り返しませんが）
workfuncのコンパイルにおいて、

```sh
make
```

を実行せずに、

```sh
make install
```

だけで良いようになったのは、2024.01の（ささやかな）改良点です。

-------------------------------------------------------

PHASE/0のコンパイル説明では、`make ... install`のみ実行しています。

https://github.com/Materials-Science-Software-Consortium/phase0_install/tree/main/WSL#phase0-%E3%82%B3%E3%83%B3%E3%83%91%E3%82%A4%E3%83%AB%E6%96%B9%E6%B3%95

**Reply by matelier** [reply]
点欠陥の生成エネルギーの計算
https://phase0-tutorial-wsl.readthedocs.io/ja/latest/defect.html
の入力ファイルでは、（He原子の計算を除き）格子定数は全て同じであることが意図されていると思います。

微妙に異なる設定が混在しているので、統一した方がわかりやすいのではないかと思い、コメントします。

```C
    unit_cell{
        #units angstrom
        a_vector = 5.98 0.00 0.00
        b_vector = 0.00 5.98 0.00
        c_vector = 0.00 0.00 5.98
    }
```

```C
    unit_cell{
        #units angstrom
        a_vector = 5.9799971975 0.00 0.00
        b_vector = 0.00 5.9799971975 0.00
        c_vector = 0.00 0.00 5.9799971975
    }
```

（AngstromとBohrの単位変換をした際の誤差、でしょうか。）

**Reply by matelier** [reply]
EV曲線計算`phase0_tutorial_wsl/wsl/Si2/ev`の`file_names.data`の擬ポテンシャル指定は、テキストと異なります。
https://phase0-tutorial-wsl.readthedocs.io/ja/latest/si.html#file-names-data-2

**Reply by matelier** [reply]
表題

    第一原理電子状態計算ソフトウェアPHASE/0 2023 チュートリアル

の`2023`は不要ではないでしょうか。もしくは`2024`に変更。
https://github.com/Materials-Science-Software-Consortium/phase0_tutorial_wsl/blob/main/source/index.rst

------------------------

`copyright`は`2024`が良いと思います。
https://github.com/Materials-Science-Software-Consortium/phase0_tutorial_wsl/blob/main/source/conf.py

**Reply by matelier** [reply]
`mpiexec`と`mpirun`は、どちらかに統一が望ましいと思います。

6.3.2. Linuxコマンド
https://phase0-tutorial-wsl.readthedocs.io/ja/latest/sup.html#linux
で、`mpiexec`が紹介されているので、こちらに統一が望ましい、でしょうか。

本文中は
2.6.1.3. シェルスクリプト
https://phase0-tutorial-wsl.readthedocs.io/ja/latest/si.html#id42
を**除き**、`mpirun`が使われているようです。
（だったら、`mpirun`に統一した方が良い、でしょうか？）

**Reply by jkoga-asms** [reply]
> `mpiexec`と`mpirun`は、どちらかに統一が望ましいと思います。

スタンダードに書かれているのは`mpiexec`らしいので，こちらに統一するのがいいと思いました

**Reply by matelier** [reply]
Si表面のldos計算のk点数は、4ではなく、15ありませんか。

https://phase0-tutorial-wsl.readthedocs.io/ja/latest/surf.html#id12

**Reply by jkoga-asms** [reply]
> Si表面のldos計算のk点数は、4ではなく、15ありませんか。

そうですね。修正します

**Comment by matelier** [comment]
# 当日反省編

## 受講者への案内

テキスト（PDF）とサンプル入力ファイルの配布手順を標準化することが望ましい。

### サンプルのZIP圧縮について

`*.zip`は扱いにくくありませんか。

WSL側で伸長する場合は、`unzip`コマンドのインストールが必要です。

Windowsで伸長してからWSL側に移動（コピー）する場合は、例えば、以下の方法を用います。

- MobaXtermのファイル転送を使うと良い；この方法を使った人は、不便を感じないでしょう。
- WSL側からLinuxコマンドで操作。`mv`コマンドを使います。`cp`の場合は、`-r`オプションが必要です。
  - [PHASE/0のインストールで説明](https://github.com/Materials-Science-Software-Consortium/phase0_install/tree/main/WSL#%E3%82%A2%E3%83%BC%E3%82%AB%E3%82%A4%E3%83%96%E3%81%AE%E3%82%B3%E3%83%94%E3%83%BC%E3%81%A8%E8%A7%A3%E5%87%8D)されているのは、オプションなしの`cp`コマンドです。
- エクスプローラーで移動すると、ファイル名に`:Zone.Identifier`を付加したファイルが余分に作成される。このファイルの存在は、`cubic.sh`の実行を妨げます。
  - ファイルの移動先によって、その所有者がrootになることがあるのかもしれない。（詳細不明）

## 実習

### 環境

延長コードがあった方が良いかもしれない。特に、参加人数が多い場合は、あった方が良いだろう。

ネットワークの提供（Wi-Fi）があった方が良いかもしれない。
追加インストールが必要な場面があり、受講者ご自身のテザリングで対応できたので問題にはならなかったが、毎回期待できるとは限らない。

### EV曲線

`bc`コマンドがインストールされていない環境が存在した。
私の手元で確認すると、Ubuntu `22.04`には`bc`コマンドなく、`24.04`には同コマンドがあった。`18.04`は未調査。
それぞれ`bc`コマンドを（直接）追加インストールした記憶はないが、インストールした何かが`bc`に依存していたために、（意識せずに）インストールされた可能性は否定できない。

Ubuntuのバージョンまで含めて（推奨）環境を指定して、事前に動作確認すれば、問題を生じない（可能性が高くなる）。

### 仕事関数

テキスト通りの動作をしない。
`workfunc.pl`が、局所ポテンシャルの平坦部を見つけることができない。

- テキストに掲載されている結果は、どのような環境で実行したものか、明記した方が良い（かもしれない）。
  - 古いバージョンや、Intelコンパイラも試したが、改善しない。
  - 本例題、最初期には擬ポテンシャルの指定が異なっており、以前の擬ポテンシャル指定で計算すると平坦部を見つけた。
    - 以前の擬ポテンシャル指定で最適化した原子配置を用いて、現在の擬ポテンシャル指定で仕事関数を計算しても、平坦部を見つけた。
    - 仕事関数の値は少し大きめ（4.89 eV）に求まった；関係ない話題かもしれない。

**Comment by jkoga-asms** [comment]
仕事関数に関しては，収束判定条件を1e-11にしたところ平坦部が見つかるようになったので，そのように修正しようとおもいます

**Reply by jkoga-asms** [reply]
> 仕事関数に関しては，収束判定条件を1e-11にしたところ平坦部が見つかるようになったので，そのように修正しようとおもいます

`sw_add_xc_to_vloc = off` を設定すると収束判定がデフォルトの状態でも平坦部が見つかるようになり，また仕事関数の値が5.05 eVとかなり大きくなります．確認はしていませんが，真空層を厚くしていくとおそらくこの値に収束していくはずなので，この方針の方がいいかもしれません

**Reply by matelier** [reply]
> 仕事関数に関しては，収束判定条件を1e-11にしたところ平坦部が見つかるようになったので，そのように修正しようとおもいます

講習会のサンプルとしてはふさわしくない、と思います。（個人の感想です）

**Comment by jkoga-asms** [comment]
EV曲線の例題については `bc` コマンドに頼らない記述の仕方を検討したいと思います

**Reply by jkoga-asms** [reply]
`bc` を使っている部分を以下のように`awk` で書き換えることができます
`awk` がインストールされていることは期待してよいでしょうか
```
   vol=$( echo | awk -v v=${v} -v dv=${dv} -v inivol=${inivol} '{print (v-1)*dv+inivol}' )
   a=$( echo | awk -v vol=${vol} '{print exp(log(vol)/3)}' )
```

**Reply by jkoga-asms** [reply]
結果を集めるために使っている`echo -n` というコマンドもポータブルではないという話があった気がします。
`echo -n`コマンドは`printf`コマンドで代替可能ですが，この書き換えでポータビリティーが向上すると考えてよいでしょうか

**Reply by matelier** [reply]
`bc`コマンドがない場合は、追加インストール、で良いと思います。

（もしくは、`python`で書き直す；`awk`の採用には反対票を投じます。）

----------------------

講習会向けには、環境を限定（ディストリビューションを指定）すれば、追加インストールが必要かどうか決まる、と期待します。

**Reply by matelier** [reply]
https://phase0-tutorial-wsl.readthedocs.io/ja/latest/si.html#id42

説明文は、`awk`を使う、に書き換えられているようです。
一方、スクリプトでは`bc`を使っているように見受けられます。

**Reply by junnara** [reply]
この記載の不一致は修正しましょう。スクリプトを修正するのではなく、説明文の記載を変えましょう（bcを使いましょう）。
環境を限定してしまうのがトラブルが一番少ないので、その方向で対応したいところです。

**Comment by jkoga-asms** [comment]
サンプルデータは `tar.gz`に圧縮しておき，トップページに置いておくと分かりやすいでしょうか

**Comment by jkoga-asms** [comment]
ReadTheDocsからPDFが作れないのは，svg変換ツールがサーバー側に
インストールされていないことが原因だと思います
ビルド時に変換ツールをインストールさせるように設定することも
可能かもしれませんが，試せていません

**Reply by jkoga-asms** [reply]
もしくはsvgはあきらめて図はすべてpngにするか

**Reply by jkoga-asms** [reply]
このコミットでsvgをpngに変更しました。その結果ReadTheDocsでpdfが作れるようになりました。
https://github.com/Materials-Science-Software-Consortium/phase0_tutorial_wsl/commit/9eec9ff8bade40345d40e17e1e1467a7346c5c44
